argoproj/argo-workflows
# fix(sso): respect `X-Forwarded-*` headers for redirect URL. Fixes #13031

Fixes #13031

### Motivation
When running Argo in plain text mode behind a TLS termination proxy with SSO enabled, you will always be redirected to the workflows page after logging in, regardless of which page you came from. This is happening because `server/auth/sso/sso.go` is using the protocol and host of the request to determine the redirect URL, which can be different than the original request when a proxy is involved.


### Modifications

The fix involves reading from the [X-Forwarded-Host](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host) and [X-Forwarded-Proto](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto) headers when present. This is almost the same as the fix that `oauth2-proxy` made when they faced a similar issue in https://github.com/oauth2-proxy/oauth2-proxy/pull/957. The only major difference is that `oauth2-proxy` only reads those headers when the `--reverse-proxy` option is set, but I didn't add a corresponding option here. The only reason I can think of for doing so is to preserve backwards-compatibility, since it's theoretically possible this could break someone's setup. I can't think of any real-world situations where that can happen, but let me know if I should put this behind a option and I'm happy to do so.


### Verification

Testing instructions:
1. Run `make start UI_SECURE=true PROFILE=sso`
2. Visit https://localhost:8080/cron-workflows
3. Click the "Login" button
4. Click "Log in with Example"
5. Click "Grant Access"
6. Confirm you're returned to https://localhost:8080/cron-workflows

https://github.com/user-attachments/assets/854d54cf-91e6-41e9-8e31-3bae20fb2722
